### ISSUE TITLE
Crash when pinging other simulation in Arm dist_bigLITTLE with SMMU

### ISSUE THREAD

**Issue by sammy17** [gem5/gem5#374]
**Describe the bug**
I was able to get arm dist_bigLITTLE setup running with SMMU as discussed in https://github.com/orgs/gem5/discussions/239
The specific changes I made are mentioned in the gem5 modifications section. 
My goal is to measure network throughput of the ethernet link between two simulations when the SMMU is present. 
Therefore, I am using timing CPU models with two Minor CPUs in the dist setup with caches enabled. 

Linux booted without any error and SMMU was recognized by Linux. 
After I setup the ethernet connections of each simulation using ifconfig, I tried to ping one simulation from the other, which caused a crash with the below error:
```
11956396690000: system.membus.snoop_filter: eraseIfNullEntry:   Removed SF entry.
11956396702000: system.realview.smmu.device_interfaces: [t] req from system.ethernet.dma addr=0xfffff000 size=0x10
11956396704000: system.realview.smmu.device_interfaces.port: micro TLB hit vaddr=0xfffff000 amask=0xfffffffffffff000 sid=0 ssid=0 paddr=0x81d90000
11956396705000: system.realview.smmu: [t] requestor req  addr=0x81d90000 size=0x10
11956396705000: system.membus.snoop_filter: lookupRequest: src system.membus.cpu_side_port[4] packet WriteReq [81d90000:81d9000f]
11956396705000: system.membus.snoop_filter: lookupRequest:   SF value 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010
11956396705000: system.bigCluster.toL2Bus.snoop_filter: lookupSnoop: packet WriteReq [81d90000:81d9000f] ES
gem5.opt: src/mem/snoop_filter.cc:225: std::pair<std::vector<gem5::QueuedResponsePort*>, gem5::Cycles> gem5::SnoopFilter::lookupSnoop(const gem5::Packet*): Assertion `cpkt->isWriteback() || cpkt->req->isUncacheable() || (cpkt->isInvalidate() == cpkt->needsWritable()) || cpkt->req->isCacheMaintenance()' failed.
Program aborted at tick 11956396705000
--- BEGIN LIBC BACKTRACE ---
```
It should be noted that without the SMMU, pinging and other network communication between two simulations work fine. 
Please refer to gem5 modifications section to see the changes made to enable SMMU in dist_bigLITTLE setup. 

**Affects version**
v23.0.1.0

**gem5 Modifications**
https://github.com/gem5/gem5/blob/48a40cf2f5182a82de360b7efa497d82e06b1631/configs/example/arm/devices.py#L461
attach_pci function was changed to add *args, **kwargs and pass **kwargs as below:
```
def attach_pci(self, dev, *args, **kwargs):
        self.realview.attachPciDevice(dev, self.iobus, **kwargs)
```
https://github.com/gem5/gem5/blob/48a40cf2f5182a82de360b7efa497d82e06b1631/configs/example/arm/dist_bigLITTLE.py#L138
addEthernet function was changed to instantiate SMMU and connect the Ethernet device to SMMU as below:
```
def addEthernet(system, options):
        # Create an SMMU but don't connect it to any devices.
        system.realview.attachSmmu([], system.membus)
        system.realview.smmu.irq_interface_enable = True
        # create NIC
        dev = IGbE_e1000()
        system.attach_pci(dev, dma_ports=[])
        system.ethernet = dev
        # Connect the device's DMA ports to the SMMU
        system.realview.smmu.connect(dev)
```
A manually edited device tree was used to include the SMMU and map the PCI ethernet device to the SMMU/IOMMU. 
Following are the relevant device tree nodes of SMMU and PCI host. 
```
	iommu: smmuv3@2b400000 {
		compatible = "arm,smmu-v3";
		reg = <0x00 0x2b400000 0x00 0x20000>;
		#iommu-cells = <0x01>;
		phandle = <0x11>;
		dma-coherent;
	};

	pci {
		compatible = "pci-host-ecam-generic";
		device_type = "pci";
		#address-cells = <0x03>;
		#size-cells = <0x02>;
		#interrupt-cells = <0x01>;
		reg = <0x00 0x40000000 0x00 0x10000000>;
		ranges = <0x1000000 0x00 0x00 0x00 0x50000000 0x00 0x10000 0x2000000 0x00 0x00 0x40 0x00 0x00 0x40000000>;
		interrupt-map = <0x00 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x44 0x01 0x800 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x45 0x01 0x1000 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x46 0x01 0x1800 0x00 0x00 0x00 0x01 0x00 0x00 0x00 0x47 0x01>;
		interrupt-map-mask = <0x1800 0x00 0x00 0x00>;
		iommu-map = <0x10 &iommu 0x0 0x1>;
		dma-coherent;
	};
```
Note that in both SMMU and pci host, dma-coherent feature is enabled. 
I kept dma-coherent for both these devices as I got to know that IO cache in gem5 handles the coherency for IO memory accesses. 
Please let me know if my understanding is incorrect. 

Also, I changed the CPU type of the BigCluster to MinorCPU to keep both systems identical https://github.com/gem5/gem5/blob/48a40cf2f5182a82de360b7efa497d82e06b1631/configs/example/arm/fs_bigLITTLE.py#L85

**To Reproduce**
Steps to reproduce the behavior. Please assume starting from a clean repository:
1. Make the above mentioned changes in devices.py and dist_bigLITTLE.py files.
2. Replace the util/dist/test/test-2nodes-AArch64.sh file with the attached file. In the attached file, I have enabled the interactive mode, caches and timing CPU model and provided the custom DTB file (which is also attached). 
3. Execute the test-2nodes-AArch64.sh to start the simulation
4. Once Linux bootup is complete, setup the ethernet in system 1 and 2 as below:
```
System 1:
/sbin/ifconfig eth0 hw ether 00:90:00:00:00:02
/sbin/ifconfig eth0 192.168.0.2 netmask 255.255.255.0 up

System 2:
/sbin/ifconfig eth0 hw ether 00:90:00:00:00:04
/sbin/ifconfig eth0 192.168.0.4 netmask 255.255.255.0 up
```
Now try to ping the system 2 from system 1:
`ping -c1 192.168.0.4`
When I execute this, simulation crash with the above-mentioned error. 

**Expected behavior**
Ping should happen without any error. 

**Host Operating System**
Ubuntu 22.04

**Host ISA**
 X86

**Compiler used**
gcc 11.3.0

**Additional information**
Linux kernel 5.15 was run on both systems with default gem5 config file. 

**test-2nodes-AArch64.sh  and dtb file that was used:**
[Archive.zip](https://github.com/gem5/gem5/files/12771011/Archive.zip)

**Comment by Harshil2107** [comment]
@giactra and @andysan do you have any idea why this would be happening?

**Comment by sammy17** [comment]
I tried to debug the issue but could not exactly figure out the root cause. 
Network communication works fine in atomic mode (mostly) without caches enabled. 
The crash only happens when the timing CPU models and caches are enabled. 
Below is a debug trace log with SMMU and Ethernet debug tracing enabled.  
Hope this helps to figure out the issue. 
Please let me know any suggestions to debug the issue further and resolve it. 
[crash.log](https://github.com/gem5/gem5/files/12812435/crash.log)

**Comment by sammy17** [comment]
Hi @giactra, any update on this? 
I would really appreciate any help in resolving this issue. 
Also, are there any other devices that are verified to work with the SMMU in gem5?

**Comment by sammy17** [comment]
I tried to debug this issue. As the error suggests, following assertion fails when performing an Ethernet packet transfer with SMMU enabled in the timing mode. 
```
// ReadEx and Writes require both invalidation and exlusivity, while reads
// require neither. Writebacks on the other hand require exclusivity but
// not the invalidation. Previously Writebacks did not generate upward
// snoops so this was never an issue. Now that Writebacks generate snoops
// we need a special case for Writebacks. Additionally cache maintenance
// operations can generate snoops as they clean and/or invalidate all
// caches down to the specified point of reference.
assert(cpkt->isWriteback() || cpkt->req->isUncacheable() ||
           (cpkt->isInvalidate() == cpkt->needsWritable()) ||
           cpkt->req->isCacheMaintenance());
```
Assert fail happen on the first WriteReq packet that arrive on lookupSnoop function. 
Following is the debug trace leading upto the crash:
```
7705626260000: system.realview.smmu: [t] requestor req  needs retry, qlen=0
7705626261000: system.realview.smmu.device_interfaces: [t] req from system.ethernet.dma addr=0xffe77b40 size=0x40
7705626261000: system.realview.smmu.device_interfaces.port: micro TLB hit vaddr=0xffe77ac0 amask=0xfffffffffffff000 sid=0 ssid=0 paddr=0x8269e000
7705626261000: system.realview.smmu.device_interfaces.port: IdHold: p=0x55ba913106e0 oid=9
7705626261000: system.realview.smmu.device_interfaces.port: IdRel:  p=0x55ba913106e0 oid=9
7705626262000: system.realview.smmu.device_interfaces: [t] req from system.ethernet.dma addr=0xffe77b80 size=0x40
7705626262000: system.realview.smmu: [t] requestor retr addr=0x8269ea40 size=0x40
7705626262000: system.membus: recvTimingReq: src system.membus.cpu_side_port[4] packet WriteReq [8269ea40:8269ea7f]
7705626262000: system.membus.snoop_filter: lookupRequest: src system.membus.cpu_side_port[4] packet WriteReq [8269ea40:8269ea7f]
7705626262000: system.membus.snoop_filter: lookupRequest:   SF value 0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010
7705626262000: system.membus: recvTimingReq: src system.membus.cpu_side_port[4] packet WriteReq [8269ea40:8269ea7f] SF size: 1 lat: 1
7705626262000: system.membus: forwardTiming for WriteReq [8269ea40:8269ea7f]
7705626262000: system.bigCluster.toL2Bus: recvTimingSnoopReq: src system.bigCluster.toL2Bus.mem_side_port[0] packet WriteReq [8269ea40:8269ea7f] ES
7705626262000: system.bigCluster.toL2Bus.snoop_filter: lookupSnoop: packet WriteReq [8269ea40:8269ea7f] ES
7705626262000: system.bigCluster.toL2Bus.snoop_filter: DEBUG: isWriteback: 0 isUncacheable: 0 isInvalidate: 0
7705626262000: system.bigCluster.toL2Bus.snoop_filter: DEBUG: needsWritable: 1 isCacheMaintenance: 0
gem5.opt: src/mem/snoop_filter.cc:237: std::pair<std::vector<gem5::QueuedResponsePort*>, gem5::Cycles> gem5::SnoopFilter::lookupSnoop(const gem5::Packet*): Assertion `cpkt->isWriteback() || cpkt->req->isUncacheable() || (cpkt->isInvalidate() == cpkt->needsWritable()) || cpkt->req->isCacheMaintenance()' failed.
```
I have printed the values that the asset checks as well. 
It seems like in the WriteReq packet, only the needsWritable attribute is set. 
From what I understood from the comment, isInvalidate attribute also should be set for this, is that correct? 
If I add the isInvalidate attribute to the WriteReq packet by changing it's definition at https://github.com/gem5/gem5/blob/e9da8d67bdbb23fbd0578a379f08f42bce50121d/src/mem/packet.cc#L79 it triggers other fundamental assertions in the abstract memory function. 

I traced back to where the WriteReq packet originate https://github.com/gem5/gem5/blob/e9da8d67bdbb23fbd0578a379f08f42bce50121d/src/dev/arm/smmu_v3_proc.cc#L122 But I am not sure how to fix the issue other than changing the packet type. 

**How can I solve this issue so that assertion does not fail? 
Can I change the packet type here?** 
It seems like WriteReq attributes contradict the comment in the assertion. 
Please let me know your thoughts. @giactra 
I would really appreciate any suggestions to solve this or debug this further.

**Comment by giactra** [comment]
Hi @sammy17 , apologies for the late reply but I am being very busy in these days and I didn't find the time to reproduce the issue.

From an initial look, it seems to me the problem is that you are attaching the SMMU directly to the membus, which is a coherent bus and therefore requires coherent transactions. This is not the case for the MemCmd::WriteReq and that's why the snoop filter complains.

You should either attach the SMMU to the non coherent xbar (iobus) and rely on the IOCache to translate it into a coherent request, or you could place a cache in front of your SMMU. Let me know if this works, otherwise I will try to allocate some time next week to look into this more carefully.

**Comment by sammy17** [comment]
Hi @giactra Thank you very much for the suggestions. 
I tried connecting the SMMU to the iobus as below. 
```
    # Following line is the only change, I added other lines for the completeness
    system.realview.attachSmmu([], system.iobus)
    # create NIC
    dev = IGbE_e1000()
    system.attach_pci(dev, dma_ports=[])
    system.ethernet = dev
    # Connect the device's DMA ports to the SMMU
    system.realview.smmu.connect(dev)
```
With this change, I ran the simulation with timing mode but during Linux bootup, the SMMU was not identified by the kernel properly. 
Linux bootup gave the following error for the SMMU:
```
[    0.032000] arm-smmu-v3 2b400000.smmuv3: unknown/unsupported TT endianness!
```
I checked the register that represent the SMMU endianess since the error suggests an issue with it. But the register is set correctly. 
And before when the SMMU is connected to the membus, Linux kernel recognized the SMMU without any issue. 
**Therefore, I believe this maybe related to a MMIO read issue when CPU reads the relevant SMMU register.** 
**Are there any other changes that I should be doing?** 
Please let me know your thoughts. 

Above message was the only message related to the SMMU in the kernel bootup, but I will add the full bootlog at the end of this message for your reference. 

I did not try the second solution you mentioned. Could you please let me know how I can connect a cache to the SMMU? 
Really appreciate any guidance on this. 

Full kernel bootup log:
```
==== m5 terminal: Terminal 0 ====
[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd070]
[    0.000000] Linux version 5.15.0 (chath@chath-g5) (aarch64-linux-gnu-gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0, GNU ld (GNU Binutils for Ubuntu) 2.38) #1 SMP PREEMPT Mon Oct 23 19:27:34 EDT 2023
[    0.000000] Machine model: V2P-CA15
[    0.000000] earlycon: pl11 at MMIO 0x000000001c090000 (options '')
[    0.000000] printk: bootconsole [pl11] enabled
[    0.000000] Memory limited to 4096MB
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000080000000-0x00000000ffffffff]
[    0.000000]   DMA32    empty
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000080000000-0x00000000ffffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000080000000-0x00000000ffffffff]
[    0.000000] cma: Reserved 16 MiB at 0x00000000fc800000
[    0.000000] percpu: Embedded 19 pages/cpu s37144 r8192 d32488 u77824
[    0.000000] pcpu-alloc: s37144 r8192 d32488 u77824 alloc=19*4096
[    0.000000] pcpu-alloc: [0] 0 [0] 1 
[    0.000000] Detected PIPT I-cache on CPU0
[    0.000000] CPU features: detected: Address authentication (architected algorithm)
[    0.000000] CPU features: detected: Spectre-v2
[    0.000000] CPU features: detected: Spectre-v4
[    0.000000] CPU features: detected: ARM erratum 834220
[    0.000000] CPU features: detected: ARM erratum 832075
[    0.000000] CPU features: detected: ARM errata 1165522, 1319367, or 1530923
[    0.000000] alternatives: patching kernel code
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 516096
[    0.000000] Kernel command line: earlyprintk earlycon=pl011,0x1c090000 console=ttyAMA0 lpj=19988480 norandmaps loglevel=8 mem=4GB root=/dev/vda1 rw init=/sbin/init vmalloc=768MB
[    0.000000] Unknown command line parameters: vmalloc=768MB
[    0.000000] Dentry cache hash table entries: 262144 (order: 9, 2097152 bytes, linear)
[    0.000000] Inode-cache hash table entries: 131072 (order: 8, 1048576 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 2027504K/2097152K available (8128K kernel code, 570K rwdata, 2752K rodata, 1280K init, 206K bss, 53264K reserved, 16384K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=2, Nodes=1
[    0.000000] rcu: Preemptible hierarchical RCU implementation.
[    0.000000] rcu: 	RCU restricting CPUs from NR_CPUS=256 to nr_cpu_ids=2.
[    0.000000] 	Trampoline variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.000000] rcu: Adjusting geometry for rcu_fanout_leaf=16, nr_cpu_ids=2
[    0.000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.000000] GICv3: GIC: Using split EOI/Deactivate mode
[    0.000000] GICv3: 988 SPIs implemented
[    0.000000] GICv3: 0 Extended SPIs implemented
[    0.000000] GICv3: Distributor has Range Selector support
[    0.000000] Root IRQ handler: gic_handle_irq
[    0.000000] GICv3: 16 PPIs implemented
[    0.000000] GICv3: CPU0: found redistributor 0 region 0:0x000000002f100000
[    0.000000] ITS: No ITS available, not enabling LPIs
[    0.000000] random: get_random_bytes called from start_kernel+0x46c/0x644 with crng_init=0
[    0.000000] arch_timer: Unable to find a suitable frame in timer @ 0x000000002a810000
[    0.000000] Failed to initialize '/timer@2a810000': -22
[    0.000000] Console: colour dummy device 80x25
[    0.000000] sched_clock: 64 bits at 250 Hz, resolution 4000000ns, wraps every 9007199254000000ns
[    0.000000] Calibrating delay loop (skipped) preset value.. 9994.24 BogoMIPS (lpj=19988480)
[    0.000000] pid_max: default: 32768 minimum: 301
[    0.000000] Mount-cache hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.000000] Mountpoint-cache hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.000000] rcu: Hierarchical SRCU implementation.
[    0.000000] smp: Bringing up secondary CPUs ...
[    0.000000] Detected PIPT I-cache on CPU1
[    0.000000] GICv3: CPU1: found redistributor 101 region 0:0x000000002f120000
[    0.000000] CPU1: Booted secondary processor 0x0000000101 [0x410fd070]
[    0.000000] smp: Brought up 1 node, 2 CPUs
[    0.000000] SMP: Total of 2 processors activated.
[    0.000000] CPU features: detected: 32-bit EL0 Support
[    0.000000] CPU features: detected: 32-bit EL1 Support
[    0.000000] CPU features: detected: Data cache clean to the PoU not required for I/D coherence
[    0.000000] CPU features: detected: Generic authentication (architected algorithm)
[    0.000000] CPU features: detected: LSE atomic instructions
[    0.000000] CPU features: detected: Privileged Access Never
[    0.000000] CPU features: detected: Scalable Vector Extension
[    0.004000] SVE: maximum available vector length 16 bytes per vector
[    0.004000] SVE: default vector length 16 bytes per vector
[    0.008000] CPU: All CPU(s) started at EL2
[    0.008000] devtmpfs: initialized
[    0.008000] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.008000] futex hash table entries: 512 (order: 3, 32768 bytes, linear)
[    0.008000] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.008000] DMA: preallocated 256 KiB GFP_KERNEL pool for atomic allocations
[    0.008000] DMA: preallocated 256 KiB GFP_KERNEL|GFP_DMA pool for atomic allocations
[    0.008000] DMA: preallocated 256 KiB GFP_KERNEL|GFP_DMA32 pool for atomic allocations
[    0.008000] cpuidle: using governor menu
[    0.012000] hw-breakpoint: found 16 breakpoint and 16 watchpoint registers.
[    0.012000] ASID allocator initialised with 256 entries
[    0.012000] Serial: AMBA PL011 UART driver
[    0.012000] OF: amba_device_add() failed (-2) for /watchdog@2a490000
[    0.012000] OF: amba_device_add() failed (-2) for /watchdog@2b060000
[    0.012000] OF: amba_device_add() failed (-2) for /kmi@1c060000
[    0.012000] OF: amba_device_add() failed (-2) for /kmi@1c070000
[    0.012000] OF: /watchdog@1c0f0000: could not find phandle 14
[    0.012000] OF: amba_device_add() failed (-2) for /watchdog@1c0f0000
[    0.012000] OF: /rtc@1c170000: could not find phandle 13
[    0.012000] OF: amba_device_add() failed (-2) for /rtc@1c170000
[    0.016000] HugeTLB registered 1.00 GiB page size, pre-allocated 0 pages
[    0.016000] HugeTLB registered 32.0 MiB page size, pre-allocated 0 pages
[    0.016000] HugeTLB registered 2.00 MiB page size, pre-allocated 0 pages
[    0.016000] HugeTLB registered 64.0 KiB page size, pre-allocated 0 pages
[    0.016000] iommu: Default domain type: Translated 
[    0.016000] iommu: DMA domain TLB invalidation policy: strict mode 
[    0.016000] vgaarb: loaded
[    0.016000] SCSI subsystem initialized
[    0.016000] libata version 3.00 loaded.
[    0.016000] usbcore: registered new interface driver usbfs
[    0.016000] usbcore: registered new interface driver hub
[    0.016000] usbcore: registered new device driver usb
[    0.016000] pps_core: LinuxPPS API ver. 1 registered
[    0.016000] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    0.016000] PTP clock support registered
[    0.020000] NET: Registered PF_INET protocol family
[    0.020000] IP idents hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.020000] tcp_listen_portaddr_hash hash table entries: 1024 (order: 2, 16384 bytes, linear)
[    0.020000] TCP established hash table entries: 16384 (order: 5, 131072 bytes, linear)
[    0.020000] TCP bind hash table entries: 16384 (order: 6, 262144 bytes, linear)
[    0.020000] TCP: Hash tables configured (established 16384 bind 16384)
[    0.020000] UDP hash table entries: 1024 (order: 3, 32768 bytes, linear)
[    0.020000] UDP-Lite hash table entries: 1024 (order: 3, 32768 bytes, linear)
[    0.020000] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.024000] RPC: Registered named UNIX socket transport module.
[    0.024000] RPC: Registered udp transport module.
[    0.024000] RPC: Registered tcp transport module.
[    0.024000] RPC: Registered tcp NFSv4.1 backchannel transport module.
[    0.024000] PCI: CLS 0 bytes, default 64
[    0.024000] kvm [1]: Guests without required CPU erratum workarounds can deadlock system!
[    0.024000] Only trusted guests should be used on this system.
[    0.024000] kvm [1]: IPA Size Limit: 40 bits
[    0.024000] kvm [1]: GICv3: no GICV resource entry
[    0.024000] kvm [1]: disabling GICv2 emulation
[    0.024000] kvm [1]: GIC system register CPU interface enabled
[    0.024000] kvm [1]: vgic interrupt IRQ9
[    0.024000] kvm [1]: kvm_arch_timer: uninitialized timecounter
[    0.024000] workingset: timestamp_bits=62 max_order=19 bucket_order=0
[    0.028000] fuse: init (API version 7.34)
[    0.028000] io scheduler mq-deadline registered
[    0.028000] io scheduler kyber registered
[    0.028000] atomic64_test: passed
[    0.028000] pci-host-generic 40000000.pci: host bridge /pci ranges:
[    0.028000] pci-host-generic 40000000.pci:   No bus range found for /pci, using [bus 00-ff]
[    0.028000] pci-host-generic 40000000.pci: Parsing ranges property...
[    0.028000] pci-host-generic 40000000.pci:       IO 0x0050000000..0x005000ffff -> 0x0000000000
[    0.028000] pci-host-generic 40000000.pci:      MEM 0x4000000000..0x403fffffff -> 0x0000000000
[    0.028000] pci-host-generic 40000000.pci: ECAM at [mem 0x40000000-0x4fffffff] for [bus 00-ff]
[    0.028000] pci-host-generic 40000000.pci: PCI host bridge to bus 0000:00
[    0.028000] pci_bus 0000:00: root bus resource [bus 00-ff]
[    0.028000] pci_bus 0000:00: root bus resource [io  0x0000-0xffff]
[    0.028000] pci_bus 0000:00: root bus resource [mem 0x4000000000-0x403fffffff] (bus address [0x00000000-0x3fffffff])
[    0.028000] pci_bus 0000:00: scanning bus
[    0.028000] pci 0000:00:01.0: [1af4:1000] type 00 class 0xff0000
[    0.028000] pci 0000:00:01.0: reg 0x10: [io  0x0000-0x001f]
[    0.028000] pci 0000:00:01.0: reg 0x30: [mem 0x4000000000-0x40000007ff pref]
[    0.028000] pci 0000:00:02.0: [8086:1075] type 00 class 0x020000
[    0.028000] pci 0000:00:02.0: reg 0x10: [mem 0x4000000000-0x400001ffff]
[    0.028000] pci 0000:00:02.0: reg 0x30: [mem 0x4000000000-0x40000007ff pref]
[    0.028000] pci_bus 0000:00: fixups for bus
[    0.028000] pci_bus 0000:00: bus scan returning with max=00
[    0.028000] pci 0000:00:02.0: BAR 0: assigned [mem 0x4000000000-0x400001ffff]
[    0.028000] pci 0000:00:01.0: BAR 6: assigned [mem 0x4000020000-0x40000207ff pref]
[    0.028000] pci 0000:00:02.0: BAR 6: assigned [mem 0x4000020800-0x4000020fff pref]
[    0.028000] pci 0000:00:01.0: BAR 0: assigned [io  0x1000-0x101f]
[    0.032000] OF: /pci: no iommu-map translation for id 0x8 on (null)
[    0.032000] virtio-pci 0000:00:01.0: assign IRQ: got 31
[    0.032000] virtio-pci 0000:00:01.0: enabling device (0000 -> 0001)
[    0.032000] virtio-pci 0000:00:01.0: virtio_pci: leaving for legacy driver
[    0.032000] virtio-pci 0000:00:01.0: enabling bus mastering
[    0.032000] Serial: 8250/16550 driver, 4 ports, IRQ sharing disabled
[    0.032000] arm-smmu-v3 2b400000.smmuv3: unknown/unsupported TT endianness!
[    0.036000] virtio_blk virtio0: [vda] 3906250 512-byte logical blocks (2.00 GB/1.86 GiB)
[    0.036000]  vda: vda1
[    0.036000] basic-mmio-gpio: Failed to locate of_node [id: -2]
[    0.036000] basic-mmio-gpio: Failed to locate of_node [id: -2]
[    0.036000] basic-mmio-gpio: Failed to locate of_node [id: -2]
[    0.116000] libphy: Fixed MDIO Bus: probed
[    0.116000] e1000: Intel(R) PRO/1000 Network Driver
[    0.116000] e1000: Copyright (c) 1999-2006 Intel Corporation.
[    0.116000] e1000e: Intel(R) PRO/1000 Network Driver
[    0.116000] e1000e: Copyright(c) 1999 - 2015 Intel Corporation.
[    0.116000] igb: Intel(R) Gigabit Ethernet Network Driver
[    0.116000] igb: Copyright (c) 2007-2014 Intel Corporation.
[    0.116000] ixgbe: Intel(R) 10 Gigabit PCI Express Network Driver
[    0.116000] ixgbe: Copyright (c) 1999-2016 Intel Corporation.
[    0.116000] ixgbevf: Intel(R) 10 Gigabit PCI Express Virtual Function Network Driver
[    0.116000] ixgbevf: Copyright (c) 2009 - 2018 Intel Corporation.
[    0.116000] i40e: Intel(R) Ethernet Connection XL710 Network Driver
[    0.116000] i40e: Copyright (c) 2013 - 2019 Intel Corporation.
[    0.116000] ixgb: Intel(R) PRO/10GbE Network Driver
[    0.116000] ixgb: Copyright (c) 1999-2008 Intel Corporation.
[    0.116000] iavf: Intel(R) Ethernet Adaptive Virtual Function Network Driver
[    0.116000] Copyright (c) 2013 - 2018 Intel Corporation.
[    0.116000] usbcore: registered new interface driver usb-storage
[    0.116000] mousedev: PS/2 mouse device common for all mice
[    0.116000] IR NEC protocol handler initialized
[    0.116000] IR RC5(x/sz) protocol handler initialized
[    0.116000] IR RC6 protocol handler initialized
[    0.116000] IR JVC protocol handler initialized
[    0.116000] IR Sony protocol handler initialized
[    0.120000] IR SANYO protocol handler initialized
[    0.120000] IR Sharp protocol handler initialized
[    0.120000] IR MCE Keyboard/mouse protocol handler initialized
[    0.120000] IR XMP protocol handler initialized
[    0.120000] usbcore: registered new interface driver usbhid
[    0.120000] usbhid: USB HID core driver
[    0.120000] NET: Registered PF_PACKET protocol family
[    0.120000] registered taskstats version 1
[    0.120000] 1c090000.uart: ttyAMA0 at MMIO 0x1c090000 (irq = 21, base_baud = 0) is a PL011 rev3
[    0.120000] printk: console [ttyAMA0] enabled
[    0.120000] printk: console [ttyAMA0] enabled
[    0.120000] printk: bootconsole [pl11] disabled
[    0.120000] printk: bootconsole [pl11] disabled
[    0.120000] 1c0a0000.uart: ttyAMA1 at MMIO 0x1c0a0000 (irq = 22, base_baud = 0) is a PL011 rev3
[    0.120000] 1c0b0000.uart: ttyAMA2 at MMIO 0x1c0b0000 (irq = 23, base_baud = 0) is a PL011 rev3
[    0.120000] 1c0c0000.uart: ttyAMA3 at MMIO 0x1c0c0000 (irq = 24, base_baud = 0) is a PL011 rev3
[    0.120000] e1000 0000:00:02.0: deferred probe timeout, ignoring dependency
[    0.120000] e1000 0000:00:02.0: assign IRQ: got 32
[    0.120000] e1000 0000:00:02.0: enabling device (0000 -> 0002)
[    0.120000] e1000 0000:00:02.0: enabling bus mastering
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x0 (reading 0x10758086)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x4 (reading 0x6)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x8 (reading 0x2000000)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0xc (reading 0x4000)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x10 (reading 0x0)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x14 (reading 0x0)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x18 (reading 0x0)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x1c (reading 0x0)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x20 (reading 0x0)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x24 (reading 0x0)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x28 (reading 0x0)
[    0.120000] e1000 0000:00:02.0: saving config space at offset 0x2c (reading 0x10088086)
[    0.124000] e1000 0000:00:02.0: saving config space at offset 0x30 (reading 0x0)
[    0.124000] e1000 0000:00:02.0: saving config space at offset 0x34 (reading 0x0)
[    0.124000] e1000 0000:00:02.0: saving config space at offset 0x38 (reading 0x0)
[    0.124000] e1000 0000:00:02.0: saving config space at offset 0x3c (reading 0xff0120)
[    0.392000] e1000 0000:00:02.0 eth0: (PCI:33MHz:32-bit) 00:90:00:00:00:01
[    0.392000] e1000 0000:00:02.0 eth0: Intel(R) PRO/1000 Network Connection
[    0.392000] EXT4-fs (vda1): mounted filesystem with ordered data mode. Opts: (null). Quota mode: disabled.
[    0.392000] VFS: Mounted root (ext4 filesystem) on device 254:1.
[    0.392000] devtmpfs: mounted
[    0.392000] Freeing unused kernel memory: 1280K
[    0.412000] Run /sbin/init as init process
[    0.412000]   with arguments:
[    0.412000]     /sbin/init
[    0.412000]   with environment:
[    0.412000]     HOME=/
[    0.412000]     TERM=linux
[    0.412000]     vmalloc=768MB
+ mount -t proc - /proc
+ mount -t sysfs - /sys
+ mount -t debugfs - /sys/kernel/debug/
+ [ -e /dev/sdb1 ]
+ [ -e /dev/sdb ]
+ [ -e /dev/vdb1 ]
+ [ -e /dev/vdb ]
+ [ -e /dev/vda1 ]
+ mount /dev/vda1 /data
+ break
+ f=/tmp/script
+ /sbin/m5 readfile
+ [ -s /tmp/script ]
+ exec /sbin/getty -a root -L ttyAMA0 vt102

Ubuntu 18.04.6 LTS aarch64-gem5 ttyAMA0
```

**Comment by sammy17** [comment]
Also, after changing the membus to iobus, when I tried to run the simulation in atomic CPU models (+ without caches), I got the following error. This issue was not seen when I run in timing mode as mentioned in the above connect, but I thought posting this as well incase it helps you to figure out the issue. 
```
src/dev/isa_fake.cc:59: warn: Device system.membus.badaddr_responder accessed by read to address 0x2b400000 size=4
src/cpu/simple/atomic.cc:405: panic: panic condition pkt.isError() occurred: Data fetch ([0x2b400000:0x2b400004]) failed: BadAddressError [2b400000:2b400003] UC
Memory Usage: 4910980 KBytes
Program aborted at tick 60713899500
```

It seems like there is an error when trying to access 0x2b400000 which is the SMMU base addr. 
Is this because caches are not available in atomic mode? 
Is the iocache enabled by default in timing mode or I have to enable it through an option?

**Comment by giactra** [comment]
> Also, after changing the membus to iobus, when I tried to run the simulation in atomic CPU models (+ without caches), I got the following error. This issue was not seen when I run in timing mode as mentioned in the above connect, but I thought posting this as well incase it helps you to figure out the issue.
> 
> ```
> src/dev/isa_fake.cc:59: warn: Device system.membus.badaddr_responder accessed by read to address 0x2b400000 size=4
> src/cpu/simple/atomic.cc:405: panic: panic condition pkt.isError() occurred: Data fetch ([0x2b400000:0x2b400004]) failed: BadAddressError [2b400000:2b400003] UC
> Memory Usage: 4910980 KBytes
> Program aborted at tick 60713899500
> ```
> 
> It seems like there is an error when trying to access 0x2b400000 which is the SMMU base addr. Is this because caches are not available in atomic mode? Is the iocache enabled by default in timing mode or I have to enable it through an option?

The fact that you get that error is logical; it’s weird that you don’t get it in timing mode…
Now that the SMMU is in the IOBus, its MMIO address range is not registered to the coherent xbar (membus). So when a request arrives, the membus doesn’t know where to dispatch it and that’s the error you are getting. You need to extend the iobridge ranges parameter to include the SMMU so that the membus will know that a request in that range should be forwarded to the bridge. 
Check the attachOnChipIO implementation to see how the bridge ranges get populated

**Comment by sammy17** [comment]
Thanks for the quick response @giactra !
It finally worked!
I added the SMMU addr range to the iobridge in the system init function where the iobridge is initialized and the SMMU was properly initialized during Linux boot. 
I can communicate between two systems without the initial crash. 
Since the initial issue is fixed, I will close this issue. Thank you very much for your help!

Unfortunately, I am facing another, independent issue for which I just filed another issue - https://github.com/gem5/gem5/issues/548. 
Hope you can take a look and help me get that resolved as well. 
Thank you in advance!
